features/dd-sdk-android-flags/src/main/kotlin/com/datadog/android/flags/internal/FlagsFeature.kt

```diff
@@ -64,6 +77,85 @@ internal class FlagsFeature(
 
     override val name: String = FLAGS_FEATURE_NAME
 
+    /**
+     * Determines if the app is running in debug mode.
+     * Uses ApplicationInfo.FLAG_DEBUGGABLE which is the proper Android way
+     * to detect if an APK was built in debug mode.
+     *
+     * Caches the result after first computation since build type doesn't change at runtime.
+     * Returns false if appContext is not yet initialized, which causes the graceful policy
+     * to be used as a safe fallback.
+     */
+    private val isDebugBuild: Boolean
+        get() {
+            // Return cached value if available (fast path)
+            debugBuildCache?.let { return it }
+
+            // Compute and cache if appContext is available
+            return appContext?.let { ctx ->
+                val isDebug = (ctx.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
+                debugBuildCache = isDebug
+                isDebug
+            } ?: run {
+                // If appContext not set yet, default to false (graceful policy) for safety.
+                // Only log once to avoid spam
+                if (debugBuildCache == null) {
+                    sdkCore.internalLogger.log(
+                        InternalLogger.Level.WARN,
+                        InternalLogger.Target.MAINTAINER,
+                        { "appContext not initialized when checking isDebugBuild - using graceful policy" }
+                    )
+                }
+                false
+            }
+        }
+
+    /**
+     * Logs an error message according to the graceful mode policy.
+     *
+     * Policy Selection:
+     * - Release builds: Graceful (log through SDK logger - conditional)
+     * - Debug + gracefulMode enabled: Error (log to System.err - always visible)
+     * - Debug + gracefulMode disabled: Strict (crash immediately)
+     *
+     * @param message The error message
+     * @param level The log level for conditional logging (graceful policy)
+     * @param shouldCrashInStrict If true, crashes in strict policy
+     */
+    internal fun logErrorWithPolicy(
+        message: String,
+        level: InternalLogger.Level = InternalLogger.Level.ERROR,
+        shouldCrashInStrict: Boolean = false
+    ) {
+        val formattedMessage = "$LOG_PREFIX $message"
+
+        when {
+            // Release build: Always graceful (conditional log through SDK)
+            !isDebugBuild -> {
+                sdkCore.internalLogger.log(
+                    level,
+                    InternalLogger.Target.USER,
+                    { formattedMessage }
+                )
+            }
+
+            // Debug + gracefulMode enabled: Error policy (always log to console)
+            flagsConfiguration.gracefulModeEnabled -> {
+                System.err.println(formattedMessage)
```

Comment:
We shouldn't use `System.log` or `System.err` for logging. We should use `FeatureSdkCore.internalLogger` instead, because:

* it is bound to the particular SDK core
* it respects verbosity set by `Datadog.setVerbosity`

Reply:
We wanted to ensure the surfacing of these misuse errors.
@ncreated - would you mind weighing in?

Reply:
Yes, Logcat (`Log` class) should be used. Although Android using Java APIs, it doesn't have a concept of `stdout` vs `stderr`, so effectively under the hood everything is redirected to Logcat when calling `System.out / System.log / System.err`.